### ISSUE TITLE
treemacs-is-file-git-ignored? does not work with symbolic links.

### ISSUE THREAD

**Issue by iboyperson** [Alexander-Miller/treemacs#218]
This is semi related to my previous https://github.com/Alexander-Miller/treemacs/issues/210 where I mentioned this but looking back to the core of that issue it deserves to stay closed. This focuses on the fact that I assume treemacs-is-file-git-ignored? is not working properly as while git status colors does show in treemacs now files in the gitignore are not hidden like they are in a non symlinked version of the same folder.

**Comment by Alexander-Miller** [comment]
Works for me, so I'll need an exact recipe of your setup and reproduction steps.

**Comment by iboyperson** [comment]
Below is the full configuration for treemacs. Additionally I installed the package from melpa if that info is needed. As far as steps to reproduce. I launch a fresh emacs and and press f8 (key to open treemacs). I then select the symlinked folder, accept the project name, and when I view it in treemacs the hidden files are shown. When I do this with the non symlinked folder in the actual path the hidden files are hidden as they should be. Below I will also include a mock up tree of the repo.

Tree:

    home
    | dir2 (symlink)
    | repo
       | .gitignore
       | dir
         | .gitignore
         | dir2
           | etc....

Config:

    (req-package treemacs
      :ensure t
      :bind
      (:map global-map
            ([f8] . treemacs)
            ("C-c f" . treemacs-select-window))
      :config
      (progn
        (setq treemacs-collapse-dirs              (if (executable-find "python") 3 0)
          treemacs-file-event-delay           5000
          treemacs-follow-after-init          t
          treemacs-follow-recenter-distance   0.1
          treemacs-goto-tag-strategy          'refetch-index
              treemacs-indentation                2
              treemacs-indentation-string         " "
              treemacs-is-never-other-window      nil
              treemacs-no-png-images              nil
              treemacs-project-follow-cleanup     nil
              treemacs-persist-file               (no-littering-expand-var-file-name "treemacs/treemacs-persist")
              treemacs-recenter-after-file-follow nil
              treemacs-recenter-after-tag-follow  nil
              treemacs-show-hidden-files          t
              treemacs-silent-filewatch           nil
              treemacs-silent-refresh             nil
              treemacs-sorting                    'alphabetic-desc
              treemacs-tag-follow-cleanup         t
              treemacs-tag-follow-delay           1.5
              treemacs-width                      35)
    
        ;; don't show files in .gitignore
        (setq treemacs-python-executable (executable-find "python3"))
        (add-to-list 'treemacs-pre-file-insert-predicates #'treemacs-is-file-git-ignored?)

        (treemacs-follow-mode t)
        (treemacs-filewatch-mode t)
        (pcase (cons (not (null (executable-find "git")))
                     (not (null (executable-find "python3"))))
          (`(t . t)
           (treemacs-git-mode 'extended))
          (`(t . _)
           (treemacs-git-mode 'simple)))))

**Comment by Alexander-Miller** [comment]
Your setup is correct. I have found the culprit - it's the very important optimization of reusing the git state fetched for the top-most directory when recursively (re-)opening directories, as it happens when you refresh or with follow-mode. 

I've now made sure to re-fetch the git status for the linked location. This might be visibly slower when you have too many symlinks around, but there's nothing I can do about that (except maybe for a deferred git-mode that may be introduced sometime in the future).

**Comment by iboyperson** [comment]
I know this is closed but the issue does not seem to have resided for me. I am using version 20180620.1257 from MELPA with Emacs version 26.1

**Comment by Alexander-Miller** [comment]
Oh boy, more edge-cases to look into!

Don't take my sarcasm the wrong way, I'll see what I can do.

**Comment by Alexander-Miller** [comment]
Are the dirs the issue happens in collapsed? In other words does it work when you set `treemacs-collapse-dirs` to 0?

**Comment by Alexander-Miller** [comment]
If your answer was yes then that gap is closed as well now.

**Comment by iboyperson** [comment]
setting treemacs-collapse-dirs to 0 does not appear to fix this. Could the issue have something to do with the .gitignore being located one directory above the symlink?

**Comment by iboyperson** [comment]
changing treemacs-collapse-dirs to 0 did not fix the issue with the latest update.  Could this issue have something to do with my .gitignore being located in the parent dir of the symlink.

**Comment by iboyperson** [comment]
Update: It appears files are being ignored just not in the root directory of the project. As an example I open a symlink to my `.emacs.d` which has `init.el` in the root of it with a `init.elc` that should be hidden but is not. However there is a directory `init.d` which contains init-treemacs.el and a init-treemacs.elc but this .elc is hidden as it should be.

Another example of this is I have a directory called `data` in the same `.emacs.d` which is gitignored. Treemacs shows the data folder and its child folders but all files within it are hidden.

**Comment by Alexander-Miller** [comment]
It's not the same as your issue, but I reproduced *something*:

Simply put: who is doing the ignoring? To get the git status of the 
files in your linked directory treemacs will start a git process inside 
that linked directory. This means that the .gitignore of the symlinked 
project will apply. So if you link one project into another it's the 
linked project that must ignore .elc files. It doesn't matter how the
top project is configured since the symlink is resolved for the purpose 
of git status gathering. I suppose this might be a source of confusion.

As for your update about link location specifics it looks like I have no 
such problems: I tried linking my dotfiles root into both my treemacs 
root and into /treemacs/src/elisp. Then I created a .elc file right 
below the dotfiles root and it's not rendered without issue.

**Comment by iboyperson** [comment]
Update: Upon looking into this further this may not be an issue with simply symlinks anymore. When attempting with my regular(non symlink) directory the same issue now pops up. Below is a screenshot of treemacs with the non-symlinked dir open in it along with my treemacs config in the other buffer. I will include the gitignore below as well. In light of this new discovery I believe this may warrant a separate issue.

.gitignore:

    # Directories
    .cache/
    config/
    data/
    elpa/

    # Files
    *.elc
    *.eld
    *~
    \#*
    \.\#*


Screenshot: Note that in both the elpa dir is expanded.
non-sym:
https://i.imgur.com/UTZztCL.png
sym:
https://i.imgur.com/OHTV0i9.png

**Comment by Alexander-Miller** [comment]
Something general is off. Without the git-ignore predicate that init.elc should have the same color as a comment. 

Let's start at the very beginning: without any symlinks, what is the output of `git status --porcelain --ignored .` when you run it in the root of your project?

**Comment by Alexander-Miller** [comment]
> Wasn't 100% sure what you wanted 

I meant doing this without any symlinks in your projects. We need to go through the basics now.

Your output looks right, elc files are clearly marked as ignored. Let's do the next steps. What the output of this:

`(treemacs-log "%s" (-> "/your/repo" (treemacs--git-status-process-function) (treemacs--git-status-parse-function)))`

The output should be a hashtable, mapping files to character numbers. 33 means ignored.

Assuming that goes well evaluate the following code:

```
(defun treemacs-is-file-git-ignored? (file git-info)
  "Determined if FILE is ignored by git by means of GIT-INFO."
  (treemacs-log "Is Ignored %s => %s" file (-some->> file (ht-get git-info) (char-to-string)))
  (eq ?! (ht-get git-info file)))
```
Then close and open any directory with a file that should be ignored. The output should tell you exactly what treemacs sees when it asks for ignored files. The output of both pieces must be directly based on git porcelain results.

**Comment by Alexander-Miller** [comment]
> The out from opening the data directory from my previous post with the code you supplied:

Here we have a bunch of hits. I assume that files like recentf-save.elc and all the other marked with `!` are correctly hidden?

> Additionally out from same code but opening .emacs.d non-symlink:

Not a single hit. I need some more detail for this. Run `trace-function` on `treemacs--git-status-process-function`, `treemacs--git-status-parse-function` and `pfuture-new` and then expand that directory again.

**Comment by Alexander-Miller** [comment]
And you can stop the tracing with `untrace-all`.

**Comment by iboyperson** [comment]
I must apologize as it appears that the treemacs-persist was interfering with previous output. Below I will include two screenshots one of the symlink and one without the symlink. This time without a treemacs-persist file causing interference.

Non-symlink:
https://i.imgur.com/VdYJ2ki.png

Symlink:
https://i.imgur.com/y7CYQu6.png

If needed I can go and rerun all the previous tests without a treemacs-persist file interfering

**Comment by Alexander-Miller** [comment]
I think the issue lies with the project root itself being a symlink. If so then that gap is closed now as well.

I should've reopened this issue a while ago now, but here's to hoping it can stay closed now.

**Comment by iboyperson** [comment]
Your hopes came true. This issue is now able to stay closed. Thank you for all the help!

**Comment by Alexander-Miller** [comment]
:fireworks:

**Comment by rileyrg** [comment]
related: follow mode is broken if using symlinks. I'm using latest treemacs from straight and emacs 29.1. I know this was an issue years ago but I thought it was fixed. Pesky symlinks ;)  An image speaks a thousand words. Here "actors" project was auto added by treemacs despite it being a sub directory of my main project at the top. Here I replaced the full ath of the parent project with its symlink shortcut.

![screenshot-latest](https://github.com/Alexander-Miller/treemacs/assets/362891/983f7611-021b-4616-8d6b-686d32b96f8c)
